File: app/src/main/kotlin/tech/mmarca/openvitals/features/heart/HeartMetricOrderedSections.kt
package tech.mmarca.openvitals.features.heart

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.runtime.Composable
import tech.mmarca.openvitals.core.period.DatePeriod
import tech.mmarca.openvitals.core.period.TimeRange
import tech.mmarca.openvitals.core.presentation.MetricDetailSectionContext
import tech.mmarca.openvitals.domain.preferences.MetricDetailSectionId
import tech.mmarca.openvitals.ui.components.MetricDetailSectionBuilder
import tech.mmarca.openvitals.ui.components.renderOrderedMetricDetailSections
import java.time.LocalDate

internal fun LazyListScope.renderHeartMetricSections(
    sectionContext: MetricDetailSectionContext,
    builder: MetricDetailSectionBuilder.() -> Unit,
) {
    renderOrderedMetricDetailSections(sectionContext, builder)
}

internal fun LazyListScope.renderChartMetricSections(
    sectionContext: MetricDetailSectionContext,
    selectedRange: TimeRange,
    period: DatePeriod,
    selectedDate: LocalDate?,
    intradayChart: (@Composable () -> Unit)? = null,
    periodChart: (@Composable () -> Unit)? = null,
    highlightCard: (@Composable () -> Unit)? = null,
    selectedDayEntries: (@Composable () -> Unit)? = null,
    dataConfidence: (@Composable () -> Unit)? = null,
    contextInsight: (@Composable () -> Unit)? = null,
    statistics: (@Composable () -> Unit)? = null,
    entries: (@Composable () -> Unit)? = null,
) {
    renderHeartMetricSections(sectionContext) {
        intradayChart?.let { chart ->
            section(MetricDetailSectionId.INTRADAY_CHART, selectedRange == TimeRange.DAY) {
                chart()
            }
        }
        periodChart?.let { chart ->
            section(MetricDetailSectionId.PERIOD_CHART, selectedRange != TimeRange.DAY || intradayChart == null) {
                chart()
            }
        }
        highlightCard?.let { card ->
            section(MetricDetailSectionId.DAILY_GOAL) {
                card()
            }
        }
        selectedDayEntries?.let { content ->
            section(MetricDetailSectionId.SELECTED_DAY_ENTRIES, selectedDate != null) {
                content()
            }
        }
        dataConfidence?.let { content ->
            section(MetricDetailSectionId.DATA_CONFIDENCE, period.start != period.end) {
                content()
            }
        }
        contextInsight?.let { content ->
            section(MetricDetailSectionId.DAILY_GOAL, highlightCard == null) {
                content()
            }
        }
        statistics?.let { content ->
            section(MetricDetailSectionId.STATISTICS) {
                content()
            }
        }
        entries?.let { content ->
            section(MetricDetailSectionId.ENTRIES) {
                content()
            }
        }
    }
}

internal fun LazyListScope.renderDayOrPeriodChartSections(
    sectionContext: MetricDetailSectionContext,
    selectedRange: TimeRange,
    period: DatePeriod,
    selectedDate: LocalDate?,
    dayChart: (@Composable () -> Unit)? = null,
    periodChart: (@Composable () -> Unit)? = null,
    highlightCard: (@Composable () -> Unit)? = null,
    selectedDayEntries: (@Composable () -> Unit)? = null,
    dataConfidence: (@Composable () -> Unit)? = null,
    contextInsight: (@Composable () -> Unit)? = null,
    statistics: (@Composable () -> Unit)? = null,
    entries: (@Composable () -> Unit)? = null,
) {
    renderChartMetricSections(
        sectionContext = sectionContext,
        selectedRange = selectedRange,
        period = period,
        selectedDate = selectedDate,
        intradayChart = dayChart,
        periodChart = periodChart,
        highlightCard = highlightCard,
        selectedDayEntries = selectedDayEntries,
        dataConfidence = dataConfidence,
        contextInsight = contextInsight,
        statistics = statistics,
        entries = entries,
    )
}

File: lib/features/heart/presentation/heart_metric_ordered_sections.dart
import 'package:flutter/material.dart';

import '../../../core/period/time_range.dart';
import '../../../core/presentation/metric_detail_sections.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/preferences/metric_detail_section_id.dart';

/// Port of the Kotlin `HeartMetricOrderedSections.kt`: the reorderable section
/// skeleton every heart + vitals metric renders through.
///
/// Visibility rules mirror `renderChartMetricSections`:
/// - the intraday chart only shows on the DAY range;
/// - the period chart shows on other ranges (or always when there is no
///   intraday chart);
/// - selected-day entries only show while a chart day is pinned;
/// - data confidence is hidden for single-day periods;
/// - the context insight only shows when there is no highlight card. Kotlin
///   registers it under `DAILY_GOAL`; here it uses `METRIC_CONTEXT` so the two
///   cannot collide in the section map and stay independently reorderable.
Widget heartChartMetricSections({
  required TimeRange selectedRange,
  required DatePeriod period,
  required LocalDate? selectedDate,
  Widget? intradayChart,
  Widget? periodChart,
  Widget? highlightCard,
  Widget? selectedDayEntries,
  Widget? dataConfidence,
  Widget? contextInsight,
  Widget? statistics,
  Widget? entries,
}) =>
    OrderedMetricDetailSections(
      sections: [
        if (intradayChart != null)
          MetricDetailSection(
            MetricDetailSectionId.intradayChart,
            visible: selectedRange == TimeRange.day,
            intradayChart,
          ),
        if (periodChart != null)
          MetricDetailSection(
            MetricDetailSectionId.periodChart,
            visible: selectedRange != TimeRange.day || intradayChart == null,
            periodChart,
          ),
        if (highlightCard != null)
          MetricDetailSection(MetricDetailSectionId.dailyGoal, highlightCard),
        if (selectedDayEntries != null)
          MetricDetailSection(
            MetricDetailSectionId.selectedDayEntries,
            visible: selectedDate != null,
            selectedDayEntries,
          ),
        if (dataConfidence != null)
          MetricDetailSection(
            MetricDetailSectionId.dataConfidence,
            visible: period.start != period.end,
            dataConfidence,
          ),
        if (contextInsight != null)
          MetricDetailSection(
            MetricDetailSectionId.metricContext,
            visible: highlightCard == null,
            contextInsight,
          ),
        if (statistics != null)
          MetricDetailSection(MetricDetailSectionId.statistics, statistics),
        if (entries != null)
          MetricDetailSection(MetricDetailSectionId.entries, entries),
      ],
    );
